streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```
@@ -1458,6 +1458,14 @@ protected boolean maybeUpdateSubscriptionVersion(final int receivedAssignmentMet
 
     @Override
     public void onAssignment(final Assignment assignment, final ConsumerGroupMetadata metadata) {
+        final Set<Task> tasksWithOpenTransactions = taskManager.allOwnedTasks()
+            .values()
+            .stream()
+            .filter(t -> t.commitNeeded())
+            .collect(Collectors.toSet());
+        log.info("Committing {} tasks with open transactions before onAssignment()", tasksWithOpenTransactions.size());
+        taskManager.commit(tasksWithOpenTransactions);
```

`taskManger.commit` might throw exception. We need to handle them. `onAssignment()` is called from within the `KafkaConsumer` which is not setup to handle any of these exceptions.

---

Hmm. I thought the exceptions propagate up to the `poll()` method, which is handled inside the `StreamThread`? Or does this blow up and mess up the rebalance handling?

---

I believe it messes up the rebalancing and might not just cleanly bubble up to `poll()`. \cc @lianetm to confirm.

I know for sure, that we have error handling logic inside `StreamsRebalanceListener#onPartitionsRevoked()`, to avoid throwing any exception from the handler back into the consumer. I assume it's the same for the partition assignor at hand.

---

I think we throw from `onPartitionsRevoked` as well. But we indeed try to finish as much as possible of the revocation before we rethrow the first exception. For KIP-1071, `handleAssignment` is called from `onPartitionsAssigned` so it will work fairly similarly.

Note also the comment below:

```
        // we do not capture any exceptions but just let the exception thrown from consumer.poll directly
        // since when stream thread captures it, either we close all tasks as dirty or we close thread
```

That sounds like it should do the right thing already?

But yes, we need to dig deep to make sure we do not break the error handling here.

---

I think the code here is right (as long as we do want to bubble up the commit failures, which I think we do). 
No matter what we throw inside the `onAssignment`, it will bubble up as a `KafkaException` on the call to `consumer.poll` (same for all rebalance callbacks). The `ConsumerCoordinator` ensures that, so no worries here.

---

Thank Lianet.

Atm, the KS code is not setup to handle an exception bubbling out of `poll()`. It would crash the whole `StreamsThreads`.

Also, if we fail here, we would not executed any of the assignment logic, so we need to make sure we don't get into corrupted internal state when we exit `onAssignment` with an error, w/o completing the assignment logic. Of course, if we let the thread crash, nothing to worry about 😂 -- but if we don't want to let `StreamsThread` just crash, we need to be very careful how to handle it.

The question is: if we cannot commit, do we really need to do an "abort everything" -- might depend on the commit error, too. In general, for EOS when a commit fails, many tasks are affected as we commit many tasks at ones, so a larger cleanup is required.